### PR TITLE
Add `image_diff` comparison method for test output verification using images

### PR DESCRIPTION
**Pull request by kostrykin**
As part of the Imaging Hackathon in Freiburg.

This PR adds the `image_diff` comparison method for test output verification using images.

Previous comparison methods comprise `diff`, `re_match`, `lines_diff`, `re_match_multiline`, `contains`, and `sim_size`, all of which are either unsuitable for image data, or dangerous to be used (because the size of a file does not really represent its content). This PR adds the `image_diff` comparison method, which is specifically tailored for single-channel and multi-channel image data (e.g., RGB), and can be used like this:

```xml
<test>
    <param name="in" value="actual.png" />
    <output name="out" value="expected.tif" compare="image_diff" metric="mae" />
</test>
```

The difference of a pair of images is quantified as the pixel-wise distance between the images, for which different *metrics* can be used. For different applications and different types of images (e.g., intensity vs. label images), different types of metrics are common. The following metrics are added for *intensity images:*

| `metric` | Corresponding metric (for intensity images) |
| --- | --- |
| `mae` | Mean absolute error |
| `mse` | Mean squared error |
| `rms` | Root mean squared error |
| `fro` | Frobenius norm |

For label images (label maps), it is usually irrelevant which particular label is used for labeling an image region, but only the *sets of pixels* corresponding to individual image regions need to be compared. This can be accomplished using *one minus* the *intersection over  the union* of the regions:

| `metric` | Corresponding metric (for label images) |
| --- | --- |
| `iou` | One minus the intersection over the union |

To decide whether a pair of images is equal, a tolerance is required, because different compression algorithms (or different versions thereof) can lead to different compression artifacts. A pair of images is considered to be *equal* in terms of the specified metric, if the distance of the images computed with respect to the metric is not above a threshold `eps` (0.01 by default).

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Review by bernt-matthias — COMMENTED** [review]
This looks really good to me. Will try to review in the next days.

If I could wish die anything, then a few linting rules for the new attributes might be cool. https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/linters/tests.py

My initial idea was more in the direction of content assertions, e.g. widht, height, colors,... but with the new dependency this should be easy. Could/should go to to a separate PR or issue.

**Comment by kostrykin** [comment]
Yes, content assertions are definitely useful too (especially powerful for label image, e.g., number of objects, average size of objects, and similar).

I plan to address this in a separate PR.

**Comment by bernt-matthias** [comment]
> @bernt-matthias Hopefully [9076175](https://github.com/galaxyproject/galaxy/commit/90761750e61fb5b84682e679ae7d459e1f7b5630) is what you meant :)

Excellent.

**Comment by nsoranzo** [comment]
> The **Test Galaxy packages** job is failing due to `ModuleNotFoundError: No module named 'numpy'`. I'm not sure, to which file do I need to add `numpy` so that it is installed as a dependency for the **Test Galaxy packages** job?

I can push a fix for this to your branch, if that's OK for you?

**Comment by kostrykin** [comment]
@nsoranzo Sure, please go ahead!

**Comment by nsoranzo** [comment]
@kostrykin I've added the changes to the package requirements, and also some type annotations if you'd like to double check them.

@jmchilton Are you OK with numpy and pillow being added as requirements to the galaxy-tool-util package?
If yes, I can also move the functions that currently use pillow conditionally (i.e. those for which I've added type annotations here) from `lib/galaxy/util/` to `tool_util/` .

File: packages/tool_util/setup.cfg [hunk 1]
```diff
@@ -36,8 +36,10 @@ install_requires =
     galaxy-util>=22.1
     conda-package-streaming
     lxml!=4.2.2
+    numpy
```
Comment (jmchilton) [inline packages/tool_util/setup.cfg]: numpy is a scary dependency and not required to parse Galaxy XML, etc.. I think we should move these to a new ``extra_require`` - maybe ``extended-assertions``.

**Comment by jmchilton** [comment]
This is really, really cool stuff - sorry about being a bit picky about the packaging stuff.

**Comment by kostrykin** [comment]
> @kostrykin I've added the changes to the package requirements, and also some type annotations if you'd like to double check them.

Thanks, the type annotations look right to me!

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
@jmchilton I've added the extra as you suggested, thanks!

**Comment by nsoranzo** [comment]
Thanks @kostrykin !

**Pull request merged**